ProvisionSample/Program.cs

```diff
@@ -579,21 +591,18 @@ static void Main(string[] args)
 
             using (var client = await CreateClient())
             {
-                // Get the newly created dataset from the previous import process
-                var datasets = await client.Datasets.GetDatasetsAsync(workspaceCollectionName, workspaceId);
-
                 // Optionally udpate the connectionstring details if preent
                 if (!string.IsNullOrWhiteSpace(connectionString))
                 {
                     var connectionParameters = new Dictionary<string, object>
                     {
                         { "connectionString", connectionString }
                     };
-                    await client.Datasets.SetAllConnectionsAsync(workspaceCollectionName, workspaceId, datasets.Value[datasets.Value.Count - 1].Id, connectionParameters);
+                    await client.Datasets.SetAllConnectionsAsync(workspaceCollectionName, workspaceId, datasetId, connectionParameters);
                 }
 
                 // Get the datasources from the dataset
-                var datasources = await client.Datasets.GetGatewayDatasourcesAsync(workspaceCollectionName, workspaceId, datasets.Value[datasets.Value.Count - 1].Id);
+                var datasources = await client.Datasets.GetGatewayDatasourcesAsync(workspaceCollectionName, workspaceId, datasetId);
```

Comment:
in this line:
await client.Gateways.PatchDatasourceAsync(workspaceCollectionName, workspaceId, datasources.Value[datasources.Value.Count - 1].GatewayId, datasources.Value[datasources.Value.Count - 1].Id, delta);
             }
         }
 
         /// <summary>
         /// Creates a new instance of the PowerBIClient with the specified token
         /// </summary>
         /// <returns></returns>
         static async Task<PowerBIClient> CreateClient()
         {
             if (accessKeys == null)
             {
                 Console.Write("Access Key: ");
                 accessKey = Console.ReadLine();
                 Console.WriteLine();
 
                 accessKeys = new WorkspaceCollectionKeys()
                 {
                     Key1 = accessKey
                 };
             }
 
             if (accessKeys == null)
             {
                 accessKeys = await ListWorkspaceCollectionKeys(subscriptionId, resourceGroup, workspaceCollectionName);
             }
 
             // Create a token credentials with "AppKey" type
             var credentials = new TokenCredentials(accessKeys.Key1, "AppKey");
 
i don't like thta we are taking the last datasource blindly. we should assert there is only one if this is the approach and take index 0 directly